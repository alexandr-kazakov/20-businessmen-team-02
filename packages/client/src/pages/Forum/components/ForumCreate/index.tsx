import React, { useCallback, useState } from 'react'
import classnames from 'classnames'

import { useAppDispatch, useAppSelector } from '../../../../app/redux/hooks'
import { getAllTopics, postTopic, setIsCreateTopic } from '../../redux/forumSlice'
import { Input } from '../../../../components/UI/Input'
import { Button } from '../../../../components/UI/Button'

import styles from './styles.module.scss'

export const ForumCreate: React.FC = () => {
  const dispatch = useAppDispatch()

  const { user } = useAppSelector(state => state.auth)

  const [values, setValues] = useState({ title: '', description: '' })
  const [isValidTitle, setIsValidTitle] = useState(true)
  const [isValidDescription, setIsValidDescription] = useState(true)

  const className = classnames(styles.description, {
    [styles.error]: !isValidDescription,
  })

  const handlerChangeTitle = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { name, value } = event.target

      if (!isValidTitle) {
        setIsValidTitle(true)
      }

      setValues({ ...values, [name]: value })
    },
    [values, isValidTitle]
  )

  const handlerChangeDescription = useCallback(
    (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      const { name, value } = event.target

      if (!isValidDescription) {
        setIsValidDescription(true)
      }

      setValues({ ...values, [name]: value })
    },
    [values, isValidDescription]
  )

  const handlerSubmit = async (event: React.FormEvent) => {
    event.preventDefault()

    if (!values.title || !values.description) {
      if (!values.title) {
        setIsValidTitle(false)
      }

      if (!values.description) {
        setIsValidDescription(false)
      }
    } else {
      const topic = {
        id_author: user?.id,
        title: values.title,
        description: values.description,
      }

      await dispatch(postTopic(topic))
      await dispatch(getAllTopics())

      dispatch(setIsCreateTopic(false))
    }
  }

  return (
    <form className={styles.form} onSubmit={handlerSubmit}>
      <div className={styles.column}>
        <Input
          className={styles.title}
          onChange={handlerChangeTitle}
          type="text"
          name="title"
          value={values.title}
          placeholder="Название"
          isValid={isValidTitle}
        />
        <textarea
          className={className}
          onChange={handlerChangeDescription}
          name="description"
          value={values.description}
          placeholder="Описание"
        />
      </div>
      <Button>Опубликовать</Button>
    </form>
  )
}
