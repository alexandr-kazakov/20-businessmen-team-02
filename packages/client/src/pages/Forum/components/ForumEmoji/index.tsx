import React from 'react'

import type { IReaction } from '../../types'

import styles from './styles.module.scss'

interface IComponent {
  emoji: IReaction
}

export const ForumEmoji: React.FC<IComponent> = props => {
  const { emoji } = props

  return <div className={styles.emoji}>{emoji.value}</div>
}
