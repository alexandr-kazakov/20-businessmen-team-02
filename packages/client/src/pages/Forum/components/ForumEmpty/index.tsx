import React from 'react'
import styles from './styles.module.scss'

interface IComponent {
  text: string
}

export const ForumEmpty: React.FC<IComponent> = props => {
  const { text } = props

  return <p className={styles.empty}>{text}</p>
}
