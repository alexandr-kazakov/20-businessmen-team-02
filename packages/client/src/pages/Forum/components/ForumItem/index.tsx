import React from 'react'

import { useAppDispatch, useAppSelector } from '../../../../app/redux/hooks'
import { getCommentsTopic, setIsCreateTopic, setSelectedIdTopic, setSelectedIdComment } from '../../redux/forumSlice'

import type { TForum } from '../../types'

import styles from './styles.module.scss'

type Props = {
  topic: TForum
}

export const ForumItem: React.FC<Props> = ({ topic }) => {
  const dispatch = useAppDispatch()

  const { isCreateTopic, selectedIdTopic, selectedIdComment } = useAppSelector(state => state.forum)

  const date = new Date(topic.createdAt).toLocaleDateString()

  const handlerClick = async () => {
    if (isCreateTopic) {
      dispatch(setIsCreateTopic(false))
    }

    if (selectedIdComment) {
      dispatch(setSelectedIdComment(null))
    }

    if (selectedIdTopic !== topic.id) {
      await dispatch(getCommentsTopic(topic.id))
      dispatch(setSelectedIdTopic(topic.id))
    }
  }

  return (
    <li className={styles.item} onClick={handlerClick}>
      <div className={styles.row}>
        <span className={styles.title}>{topic.title}</span>
        <time className={styles.date}>{date}</time>
      </div>
      <p className={styles.description}>{topic.description}</p>
    </li>
  )
}
