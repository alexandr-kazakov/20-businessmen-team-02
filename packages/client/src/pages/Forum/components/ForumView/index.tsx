import React from 'react'
import { useAppSelector } from '../../../../app/redux/hooks'
import styles from './styles.module.scss'

export const ForumView: React.FC = () => {
  const { selectedForum } = useAppSelector(state => state.forum)

  return (
    <div className={styles.view}>
      {selectedForum ? (
        <>
          <div className={styles.row}>
            <span className={styles.title}>{selectedForum.title}</span>
            <time className={styles.date}>{selectedForum.date}</time>
          </div>
          <p className={styles.description}>{selectedForum.description}</p>
        </>
      ) : (
        <>
          <p className={styles.description}>Выберите форум</p>
        </>
      )}
    </div>
  )
}
