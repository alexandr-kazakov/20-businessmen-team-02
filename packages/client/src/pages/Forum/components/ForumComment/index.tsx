import React, { useState } from 'react'
import Picker from '@emoji-mart/react'

import { useAppDispatch, useAppSelector } from '../../../../app/redux/hooks'
import { getComments, createReaction } from '../../redux/forumSlice'
import { ForumEmoji } from '../ForumEmoji'

import type { IComment } from '../../types'

import styles from './styles.module.scss'

interface IComponent {
  comment: IComment
}

export const ForumComment: React.FC<IComponent> = props => {
  const { comment } = props

  const dispatch = useAppDispatch()

  const { user } = useAppSelector(state => state.auth)

  const [isEmoji, setIsEmoji] = useState(false)

  const toggleEmoji = () => {
    setIsEmoji(prev => !prev)
  }

  const onEmojiSelect = async (event: any) => {
    const reaction = {
      id_comment: comment.id,
      id_author: String(user?.id),
      value: event.native,
    }

    await dispatch(createReaction(reaction))
    await dispatch(getComments(comment.id_topic))

    toggleEmoji()
  }

  return (
    <div className={styles.comment}>
      <div className={styles.container}>
        <div className={styles.wrapper}>
          <span className={styles.author}>{comment.login_author}</span>
        </div>
        <p className={styles.text}>{comment.text}</p>
      </div>
      <div className={styles.reactions}>
        <button className={styles.plus} onClick={toggleEmoji}>
          +
        </button>
        {comment.Reactions.length !== 0 && comment.Reactions.map(emoji => <ForumEmoji key={emoji.id} emoji={emoji} />)}
        {isEmoji && (
          <div className={styles.picker}>
            <Picker
              onEmojiSelect={onEmojiSelect}
              categories={['frequent']}
              searchPosition="none"
              navPosition="none"
              previewPosition="none"
              emojiButtonSize="20"
              emojiSize="16"
            />
          </div>
        )}
      </div>
    </div>
  )
}
