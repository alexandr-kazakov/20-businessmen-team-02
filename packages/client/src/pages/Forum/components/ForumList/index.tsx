import React from 'react'
import { useAppSelector } from '../../../../app/redux/hooks'
import { ForumItem } from '../ForumItem'
import styles from './styles.module.scss'

export const ForumList: React.FC = () => {
  const { listForums } = useAppSelector(state => state.forum)

  return (
    <ul className={styles.list}>
      {listForums.length !== 0 ? (
        <>
          {listForums.map(forum => (
            <ForumItem key={forum.id} forum={forum} />
          ))}
        </>
      ) : (
        <p className={styles.empty}>Список пуст</p>
      )}
    </ul>
  )
}
