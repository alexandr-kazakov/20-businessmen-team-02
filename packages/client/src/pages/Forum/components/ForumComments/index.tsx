import React from 'react'

import { useAppSelector } from '../../../../app/redux/hooks'
import { ForumComment } from '../ForumComment'
import { ForumEmpty } from '../ForumEmpty'

import styles from './styles.module.scss'

export const ForumComments: React.FC = () => {
  const { commentsTopic, selectedComment } = useAppSelector(state => state.forum)

  return (
    <div className={styles.comments}>
      {selectedComment ? (
        <>
          <ForumComment comment={selectedComment} />
          <div className={styles.answers}>
            {selectedComment.Comments.length === 0 ? (
              <div className={styles.wrapper}>
                <ForumEmpty text="Ответы на этот коммент отсутствуют" />
              </div>
            ) : (
              <div className={styles.container}>
                {selectedComment.Comments.map((comment: any) => (
                  <ForumComment key={comment.id} comment={comment} />
                ))}
              </div>
            )}
          </div>
        </>
      ) : (
        <>
          {commentsTopic.length === 0 ? (
            <ForumEmpty text="Комментов нет" />
          ) : (
            commentsTopic.map((comment: any) => <ForumComment key={comment.id} comment={comment} />)
          )}
        </>
      )}
    </div>
  )
}
