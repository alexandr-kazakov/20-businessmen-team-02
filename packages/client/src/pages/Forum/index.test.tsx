import React from 'react'
import { render } from '@testing-library/react'

import { ForumList } from '@/pages/Forum/components/ForumList'
import { ForumView } from '@/pages/Forum/components/ForumView'

jest.mock('react-redux', () => ({
  useSelector: () => ({ isCreateTopic: true, topicsList: [] }),
  useDispatch: jest.fn(),
}))

describe('Forum', () => {
  test('should render ForumList component', () => {
    const { asFragment } = render(<ForumList />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render ForumView component', () => {
    const { asFragment } = render(<ForumView />)

    expect(asFragment()).toMatchSnapshot()
  })
})
