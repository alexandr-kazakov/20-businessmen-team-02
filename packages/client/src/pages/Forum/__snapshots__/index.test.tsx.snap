// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forum should render ForumList component 1`] = `
<DocumentFragment>
  <div>
    <div>
      <button
        class=""
      >
        Создать топик
      </button>
    </div>
    <ul>
      <p>
        Список пуст
      </p>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Forum should render ForumView component 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <input
          class=""
          name="title"
          placeholder="Название"
          type="text"
          value=""
        />
        <textarea
          class=""
          name="description"
          placeholder="Описание"
        />
      </div>
      <button
        class=""
      >
        Опубликовать
      </button>
    </form>
  </div>
</DocumentFragment>
`;
